### PR TITLE
client: accommodate Consul 1.14.0 gRPC and agent self changes.

### LINKED ISSUE #15266: Consul grpc_tls usage cannot deploy nomad jobs anymore
### Nomad version
1.4.2
### Consul version
1.14.0

I updated my consul agent today. Consul 1.40 have a breaking change 

BREAKING CHANGES:
config: Add new ports.grpc_tls configuration option.

I remove the grpc falg form the config and replace it with grpc_tls . After that nomad is not able to deploy jobs with the error:

Placement Failure
Task Group "keycloak":
  * Constraint "${attr.consul.grpc} > 0": 1 nodes excluded by filter

### PR DESCRIPTION
**Pull request by jrasell**
Consul 1.14.0 changed the way in which gRPC listeners are configured, particularly when using TLS. Prior to the change, a single listener was responsible for handling plain-text and encrypted gRPC requests. In 1.14.0 and beyond, separate listeners will be used for each, defaulting to 8502 and 8503 for plain-text and TLS respectively.

The change means that Nomad’s Consul Connect integration would not work when integrated with Consul clusters using TLS and running 1.14.0 or greater.

The Nomad Consul fingerprinter identifies the gRPC port Consul has exposed using the "DebugConfig.GRPCPort" value from Consul’s “/v1/agent/self” endpoint. In Consul 1.14.0 and greater, this only represents the plain-text gRPC port which is likely to be disbaled in clusters running TLS. In order to fix this issue, Nomad now takes into account the Consul version and configured scheme to optionally use “DebugConfig.GRPCTLSPort” value from Consul’s agent self return.

The “consul_grcp_socket” allocrunner hook has also been updated so that the fingerprinted gRPC port attribute is passed in. This provides a better fallback method, when the operator does not configure the “consul.grpc_address” option.

Closes #15266

### CODE REVIEW

File: client/fingerprint/consul.go [hunk 1]
```diff
@@ -173,11 +183,54 @@ func (f *ConsulFingerprint) connect(info agentconsul.Self) (string, bool) {
...
+		}
+
+		consulVersion, err := version.NewVersion(v)
+		if err != nil {
+			return "", false
+		}
+
+		// This check ensures that users running a 1.14.0 pre-release get the
+		// desired behaviour, otherwise 1.14.0-beta1 < 1.14.0 will be truthy.
+		versionsMatch := slices.Equal(consulGRPCPortChangeVersion.Segments(), consulVersion.Segments())
```
Comment (tgross) [inline client/fingerprint/consul.go:204]: We probably don't use it everywhere we should be but just FYI [`Version.Core`](https://pkg.go.dev/github.com/hashicorp/go-version?utm_source=godoc#Version.Core) lets you extract a version without `-beta1`, etc.

File: client/fingerprint/consul_test.go [hunk 2]
```diff
@@ -296,29 +296,127 @@ func TestConsulFingerprint_grpc(t *testing.T) {
...
+	})
+
+	t.Run("grpc disabled pre-1.14 https", func(t *testing.T) {
+		s, ok := fp.grpc("https")(agentconsul.Self{
+			"Config":      {"Version": "1.14.0"},
+			"DebugConfig": {"GRPCTLSPort": -1.0}, // JSON numbers are floats
+		})
+		require.True(t, ok)
+		require.Equal(t, "-1", s)
+	})
```
Comment (tgross) [inline client/fingerprint/consul_test.go:369]: This test is repeated below on line 398. But it also looks weird to me: it says "pre-1.14" but the version is 1.14.0 and I wouldn't expect a pre-1.14 to have the `GRPCTLSPort` field at all? Maybe just the name is wrong?

File: client/fingerprint/consul.go (hunk 1 above)
Comment (jrasell) [reply client/fingerprint/consul.go]: Thanks that is a lot cleaner. The update can be seen within commit: https://github.com/hashicorp/nomad/pull/15309/commits/cebd6b59ed3afbc521678cb3441be1c2ce5b64d2

File: client/fingerprint/consul_test.go (hunk 2 above)
Comment (jrasell) [reply client/fingerprint/consul_test.go]: I think I had a mare with copy/pasting the test outlines; this has been cleaned up within commit: https://github.com/hashicorp/nomad/pull/15309/commits/cebd6b59ed3afbc521678cb3441be1c2ce5b64d2

**Review by tgross — APPROVED** [review]
LGTM

File: client/fingerprint/consul.go [hunk 3]
```diff
@@ -173,11 +182,50 @@ func (f *ConsulFingerprint) connect(info agentconsul.Self) (string, bool) {
...
+		// than 1.14.0 we use the original single gRPC port.
+		if consulVersion.Core().LessThan(consulGRPCPortChangeVersion.Core()) {
+			return f.grpcPort(info)
+		}
+
+		// Now that we know we are querying a Consul agent running v1.14.0 or
+		// greater, we need to select the correct port parameter from the
+		// config depending on whether we have been asked to speak TLS or not.
+		switch scheme {
+		case "https", "HTTPS":
```
Comment (schmichael) [inline client/fingerprint/consul.go:211]: Not a blocker, but I think lowering (or raising) is a better approach generally:

```suggestion
		switch strings.ToLower(scheme) {
		case "https":
```

**Review by schmichael — APPROVED** [review]
Thanks for jumping on this @jrasell! LGTM

File: client/fingerprint/consul.go (hunk 3 above)
Comment (jrasell) [reply client/fingerprint/consul.go]: Agreed; I was trying to be clever and avoid the func overhead but I can now see what can go wrong with that approach. This has been fixed in commit: https://github.com/hashicorp/nomad/pull/15309/commits/497eb98b161768cb8ccc54e9a3f577d90908470e

**Review by shoenig — APPROVED** [review]
Thanks @jrasell !

**Pull request merged**